Account/FirefoxAccount.swift

```diff
@@ -190,13 +196,104 @@ open class FirefoxAccount {
             return "Not in a Token State: \(state?.label.rawValue ?? "Empty State")"
         }
     }
-
+    
+    public class FxAProfile {
+        open var displayName: String?
+        open let email: String
+        open let avatar: Avatar
+        
+        init(email: String, displayName: String?, avatar: String?) {
+            self.email = email
+            self.displayName = displayName
+            self.avatar = Avatar(url: avatar?.asURL)
+        }
+        
+        enum ImageDownloadState {
+            case notStarted
+            case started
+            case failedCanRetry
+            case failedCanNotRetry
+            case succeededMalformed
+            case succeeded
+        }
+        
+        open class Avatar {
+            open var image: UIImage?
+            open let url: URL?
+            var currentImageState: ImageDownloadState = ImageDownloadState.notStarted
+            
+            init(url: URL?) {
+                self.image = UIImage(named: "placeholder-avatar")
+                self.url = url
+                self.updateAvatarImageState()
+            }
+            
+            func updateAvatarImageState() {
+                switch currentImageState {
+                case .notStarted:
+                    self.currentImageState = ImageDownloadState.started
+                    self.downloadAvatar()
+                    break
+                case .failedCanRetry:
+                    self.downloadAvatar()
+                    break
+                default:
+                    break
+                }
+            }
+            
+            func downloadAvatar() {
+                SDWebImageManager.shared().downloadImage(with: url, options: [SDWebImageOptions.continueInBackground, SDWebImageOptions.lowPriority], progress: nil) { (image, error, cacheType, success, url) in
+                    if let error = error {
+                        if (error as NSError).code == 404 || self.currentImageState == ImageDownloadState.failedCanRetry {
+                            // Image is not found or failed to download a second time
+                            self.currentImageState = ImageDownloadState.failedCanNotRetry
+                        } else {
+                            // This could have been a transient error, attempt to download the image only once more
+                            self.currentImageState = ImageDownloadState.failedCanRetry
+                            self.updateAvatarImageState()
+                        }
+                        return
+                    }
+                    
+                    if success == true && image == nil {
+                        self.currentImageState = ImageDownloadState.succeededMalformed
```

Comment:
This might occur due to a captive portal… let's make sure that a user on Starbucks' wifi doesn't end up with a permanent missing avatar. Or does this all reset on next launch?

Reply:
This would reset upon next launch, so the avatar would have another chance to load.